[FEATURE] timetable context 생성

goranikin opened the pull request.

goranikin committed:
context 파일 생성

goranikin committed:
timetable context 생성 완료!

goranikin committed:
lint 불 끄기

Yeonu-Kim commented on src/App.tsx:
```diff
@@ -221,6 +222,24 @@ export const App = () => {
 
+  // timetableId 저장 함수를 context api로 관리
+  const [timetableId, setTimetableId] = useState<string | null>(() => {
+    const savedTimetableId = localStorage.getItem('selectedTimetableId');
```
localstorage에서 가지고 오는 건 repository에서 관리해야 할 거 같아요

Yeonu-Kim replied:
https://github.com/Yeonu-Kim/snutt_clone_7/blob/main/src/infrastructure/impleStorageRepository.ts

Yeonu-Kim commented on src/App.tsx:
```diff
@@ -221,6 +222,24 @@ export const App = () => {
+
+    if (savedTimetableId === null) return null;
+
+    return JSON.parse(savedTimetableId) as string;
```
혹시 뒤에 as string은 왜 붙였죠..?? 이렇게 타입 우기기를 하면 안될 거 같고, 그냥 savedTimetableId를 return하면 될 거 같아용

Yeonu-Kim commented on src/App.tsx:
```diff
@@ -221,6 +222,24 @@ export const App = () => {
+  const updateTimetableId = (id: string | null) => {
+    setTimetableId(id);
+    if (id !== null) {
+      localStorage.setItem('selectedTimetableId', JSON.stringify(id));
```
id 자체가 string인데 굳이 stringify를 해야 하나요?

goranikin replied:
아항 글쿠만요

Yeonu-Kim reviewed (approved):
localstorage 사용해서 관리하고 있는데, token service를 어떻게 관리하고 있는지 확인해보고 그거랑 비슷하게 작업하시면 될 거 같습니다!

goranikin committed:
stringfy 없애고 as string 없애기

goranikin committed:
storage에 저장된 Id를 초기화해줘야 하는데...

goranikin committed:
완성!
현재 선택한 시간표를 삭제하면 queryClient로 다시 쿼리를 보내 getTimetableById가 아닌 getTimetable로 다시 가져와 local storage로 됨!

goranikin committed:
lint

Yeonu-Kim commented on src/App.tsx:
```diff
@@ -221,6 +229,11 @@ export const App = () => {
   };
 
+  // timetableId 저장 함수를 context api로 관리
+  const [timetableId, setTimetableId] = useState(
+    timetableStorageRepository.getStorageTimetableId,
+  );
```
마이너하긴 한데, token state 바로 아래에 위치시키면 좋을 듯 합니다! state 끼리 모아져있는게 보기 편해서요

goranikin replied:
반영했어요

Yeonu-Kim commented on src/pages/Main/Drawer/TimeTableMenuBottomSheet/DeleteDialog.tsx:
```diff
@@ -17,7 +17,7 @@ export const DeleteDialog = ({
   timetableId: string;
   selectedTimetableId: string | null;
-  setTimetableId: React.Dispatch<React.SetStateAction<string | null>>;
+  setTimetableId: (timetableId: string | null) => void;
```
setTimetableId가 이미 전역으로 올라가 있으니까 굳이 setTimetableId를 props로 받아야 하나 싶네요.

Yeonu-Kim replied:
그리고 아래 67line에서 setTimetableId 실행 시 state는 바뀌지만 localstorage에 저장된 값은 바뀌지 않을 건데요.
 timetableService에서 추가하신 메서드로 localstorage 값을 관리하면 될 거 같습니다.

Yeonu-Kim commented on src/pages/Main/Drawer/index.tsx:
```diff
@@ -18,7 +18,7 @@ type Drawer = {
   onClose: () => void;
   selectedTimetableId: string | null;
-  setTimetableId: React.Dispatch<React.SetStateAction<string | null>>;
+  setTimetableId: (timetableId: string | null) => void;
```
위 리뷰와 마찬가지로
1. 굳이 props로 안 받아도 되고
2. localstorage 값 변경해줘야 할 거 같습니다.

goranikin replied:
```
asda
```

Yeonu-Kim commented on src/pages/Main/index.tsx:
```diff
@@ -11,10 +14,18 @@ import { TimeTable } from './TimeTable';
   const [totalCredit, setTotalCredit] = useState(0);
   const [title, setTitle] = useState('');
   const { toLectureList } = useRouteNavigation();
+  const { timetableId, setTimetableId } = useGuardContext(TimetableContext);
+  const { timeTableService } = useGuardContext(ServiceContext);
+
+  const handleClickSetTimetableId = (selectedTimetableId: string | null) => {
+    setTimetableId(selectedTimetableId);
+    timeTableService.storeSelectedTimetableId({
+      selectedTimetableId: selectedTimetableId,
+    });
+  };
```
위 리뷰와 마찬가지로 상위에서 불러와서 props로 넘기는 거 자체가 불필요할 듯 합니다.

goranikin replied:
props로 넘기는 거에 대해 이름을 오해하게 썼는데,
```
setTimetableId={handleClickSetTimetableId}
```
라는 방식으로 Drawer랑 Timatable 컴포넌트에 props로 넘기고 있어요.
```
handleClickSetTimetableId = {동일}
```
로 바꾸겠습니다.

그리고 handleClickSetTimetableId를 넘기고 하위 컴포넌트에서 setTimetableId의 이름으로 사용하고 있긴 하지만, 기능을 보면
```
setTimetableId(selectedTimetableId);
timetableService.storeSelectedTimetableId(
      selectedTimetableId: selectedTimetableId,
    });
  };
```
로 작성되어 있어 LocalStorage는 잘 바뀝니다.

그런데 setTimetableId와 TimetableService는 context api로 전역에 올리긴 했지만, Main에서 각각의 두 컨텍스트를 이용해 handler function을 만들어줬다면 얘는 그냥 props로 넘기는 게 중복을 제거하는 면에서 좋다고 생각했어요.

Yeonu-Kim replied:
아하 그렇군요?? 그럼 좋은 거 같아요

goranikin replied:
사실 이렇게 관리해보면서 느낀 건데, Service에서 LocalStorage를 따로 관리하는 메서드를 담당하고, 또 내부적으로 state를 관리하는 걸 따로 두는 이유가 뭔가요. 이 둘을 합치는 게 서버와의 데이터와 내부 스테이트의 동기를 더 잘 해줄 수 있을 거 같은데.

Yeonu-Kim replied:
state를 사용하는 이유 -> timetableId가 바뀌었을 때 해당 컴포넌트가 리렌더되도록 하기 위함.
localstorage를 사용하는 이유 -> 페이지가 꺼져도 저장하기 위함.

두개를 합치는 건 저도 고민해봤는데, 구조상 안되더라고요. 만약에 TimetableIdContext가 있다고 치면 그 안에 ServiceContext 안에 있는 것들을 넣게되는 거잖아요?? 그래서 분리하는 듯 합니다. (시도를 안해봐서 확실하지는 X)

goranikin committed:
timetableId state 위치 이동

goranikin committed:
setTimetableId라 헷갈리는 이름을 handleClicksetTimetableId로 바꿧음

goranikin committed:
이름 안 바꾼 거 fix
